### PR TITLE
Quest.PirateRealm

### PR DESCRIPTION
**Pull request by gausie**
The goal here is to create a Quest in the KoLmafia format for a journey through PirateRealm. Thanks to @Shiverwarp for helping me flesh out a step plan.

<details>
  <summary>Step plan hidden here for brevity (click to view)</summary>

  ```
 unstarted: We have not yet acquired our eyepatch
 started: We have acquired our eyepatch, but We have not yet finished choosing all our "start pirate realm" things (ship, crew, curio)
 step 1: We've chosen all our starting equipment, but have not yet chosen our first island
  
  // Island 1
 step 2: We have chosen our first island, We are in the process of sailing to the first island (sailing turns are less than our ship speed)
 step 3: We have finished our sailing turns for the first island, but we have not yet seen the Land Ho NC for the first island
 step 4: We have seen  the land ho NC for first island, and are in the process of doing the 4 combats on the first island
 step 5: we have completed the 4 combats, and have the final encounter of the first island available
 step 6: we have completed the final encounter of the first island, but have not yet chosen our second island
  
  // Island 2
  step 7: we have chosen our second island, we are in the process of sailing to the second island (sailing turns are less than our ship speed)
  step 8: We have finished our sailing turns for the second island, but we have not yet seen the Land Ho NC for the second island
  step 9: We have seen the land ho NC for second island, and are in the process of doing the 4 combats on the second island
  step 10: we have completed the 4 combats on the second island, and have the final encounter of the second island available
  step 11: we have completed the final encounter of the second island, but have not yet chosen our third and final island
  
  // Island 3
  step 12: we have chosen our third and final island, we are in the process of sailing to the third island (sailing turns are less than our ship speed)
  step 13: We have finished our sailing turns for the third island, but we have not yet seen the Land Ho NC for the third island
  step 14: We have seen the land ho NC for final island, and are in the process of doing the 9 combats on the second island
  step 15: we have completed the 9 combats on the final island, and have the final encounter of the third and final island available
  step 16: we have completed the final encounter of the final island, we have not yet gotten our funalog
  
  finished: piraterealm complete for the day
  ```
</details>

In the course of writing this I have realised that Quests are usually exclusively reserved for things that appear in the quest log. This doesn't and I've had to work around that by adding a fake entry to the questlogs.txt datafile. Do we think that is ok? Using the Quest progress utilities is so useful here.

### CODE REVIEW

File: src/net/sourceforge/kolmafia/request/FightRequest.java [hunk 1]
```diff
@@ -4216,7 +4216,10 @@ else if (responseText.contains("a horrible grinning clown head emerges")) {
 
       // Track monsters defeated on the current PirateRealm island
       if (adventure == AdventurePool.PIRATEREALM_ISLAND) {
-        Preferences.increment("_pirateRealmIslandMonstersDefeated");
+        var defeated = Preferences.increment("_pirateRealmIslandMonstersDefeated");
+        if (defeated == (QuestManager.getPirateRealmIslandNumber() < 2 ? 4 : 9)) {
+          QuestManager.setPirateRealmIslandQuestProgress(3);
```
Comment (midgleyc) [inline src/net/sourceforge/kolmafia/request/FightRequest.java:4221]: should also update quest progress when windicle used?

File: src/net/sourceforge/kolmafia/session/ChoiceControl.java [hunk 2]
```diff
@@ -424,6 +424,30 @@ public static void postChoice0(int choice, final String urlString, final Generic
...
+          // Advance quest progress
+          QuestManager.setPirateRealmIslandQuestProgress(choice - 1352, 0);
+          // Reset per-island flags
+          Preferences.setInteger("_pirateRealmIslandMonstersDefeated", 0);
+          Preferences.setInteger("_pirateRealmSailingTurns", 0);
+          Preferences.setBoolean("_pirateRealmWindicleUsed", false);
+          break;
+        }
+
+      case 1355:
```
Comment (midgleyc) [inline src/net/sourceforge/kolmafia/session/ChoiceControl.java:444]: please add a comment with the adventure name, like the other choices

File: src/net/sourceforge/kolmafia/session/ChoiceControl.java [hunk 3]
```diff
@@ -424,6 +424,30 @@ public static void postChoice0(int choice, final String urlString, final Generic
         BastilleBattalionManager.preChoice(urlString, request);
         break;
 
+      case 1350:
```
Comment (midgleyc) [inline src/net/sourceforge/kolmafia/session/ChoiceControl.java:427]: please add a comment with the adventure name, like the other choices

File: src/net/sourceforge/kolmafia/session/QuestManager.java [hunk 4]
```diff
@@ -795,13 +795,39 @@ private static void addFantasyRealmKill(String monster) {
...
+    return 2;
+  }
+
+  public static void setPirateRealmIslandQuestProgress(final int progress) {
+    var island = getPirateRealmIslandNumber();
+    setPirateRealmIslandQuestProgress(island, progress);
+  }
+
+  public static void setPirateRealmIslandQuestProgress(final int island, final int progress) {
+    QuestDatabase.setQuestIfBetter(Quest.PIRATEREALM, (island * 5) + 2 + progress);
```
[the remark this answers is not in the document]
Comment (gausie) [reply src/net/sourceforge/kolmafia/session/QuestManager.java:815]: I'll eventually put it in the mafia wiki, but it's in the collapsed section in the description for this PR

[the remark this answers is not in the document]
Comment (midgleyc) [reply src/net/sourceforge/kolmafia/session/QuestManager.java:815]: step 6 and step 11 appear to never be set?

[the remark this answers is not in the document]
Comment (gausie) [reply src/net/sourceforge/kolmafia/session/QuestManager.java:815]: Huh. I wrote code for that. I need to go through my stash stack.

**Review by midgleyc — APPROVED** [review]
OK!

**Comment by gausie** [comment]
Did my runthrough today and there are a couple of fixes/recovery states I want to add. I am getting some "Bad pipe" errors today for some reason which causes things to get out of sync, and there are some quest progression steps that can be recovered just from looking at the place.php

**Pull request merged**